rebalance/rebalance.py

```diff
@@ -212,16 +214,260 @@ def rebalance(plugin, outgoing_scid, incoming_scid, msatoshi: Millisatoshi=None,
+    i = 0
+    while True:
+        amount1 = min(must_send(ch1, enough_liquidity), could_receive(ch2, enough_liquidity))
+        amount2 = min(should_send(ch1, enough_liquidity, ideal_ratio), should_receive(ch2, enough_liquidity, ideal_ratio))
+        amount3 = min(could_send(ch1, enough_liquidity), must_receive(ch2, enough_liquidity))
+        amount = max(amount1, amount2, amount3)
+        if amount < min_amount:
+            return result
+        amount = int(min(amount, get_max_amount(i, min_amount, enough_liquidity)))
+        maxfee = get_max_fee(plugin, amount)
+        plugin.log(f"Try to rebalance: {scid1} -> {scid2}; amount={amount:,} msat; maxfee={maxfee:,} msat")
+        try:
+            res = rebalance(plugin, outgoing_scid=scid1, incoming_scid=scid2, msatoshi=amount, maxfeepercent=0, retry_for=1200, exemptfee=maxfee)
+        except Exception:
+            failed_pairs.append(scid1 + ":" + scid2)
+            # rebalance failed, let's try with a smaller amount
+            while (get_max_amount(i, min_amount, enough_liquidity) >= amount and
+                   get_max_amount(i, min_amount, enough_liquidity) != get_max_amount(i + i, min_amount, enough_liquidity)):
+                i += 1
+            if amount > get_max_amount(i, min_amount, enough_liquidity):
+                continue
+            return result
+        plugin.log(f"Rebalance succeeded: {res}")
+        result["success"] = True
+        result["fee_spent"] += res["fee"]
+        # refresh channels
+        time.sleep(10)
+        ch1 = get_chan(plugin, scid1)
+        assert ch1 is not None
+        ch2 = get_chan(plugin, scid2)
+        assert ch2 is not None
+
+
+def maybe_rebalance_once(plugin: Plugin, failed_pairs: list):
+    channels = get_our_channels(plugin)
+    for ch1 in channels:
+        for ch2 in channels:
+            result = maybe_rebalance_pairs(plugin, ch1, ch2, failed_pairs)
+            if result["success"] == True:
+                return result
+    return {"success": False, "fee_spent": 0}
+
+
+def rebalanceall_thread(plugin: Plugin):
+    try:
+        channels = get_our_channels(plugin)
+        enough_liquidity = get_enough_liquidity_threshold(channels)
+        ideal_ratio = get_ideal_ratio(channels, enough_liquidity)
+        plugin.log(f"Automatic rebalance is running with enough liquidity threshold: {enough_liquidity:,} msat, "
+                   f"ideal liquidity ratio: {ideal_ratio * 100:.2f}%, "
+                   f"min rebalancable amount: {int(Millisatoshi(plugin.get_option('min-amount'))):,} msat, "
+                   f"feeratio: {plugin.get_option('feeratio')}")
+        plugin.options["enough-liquidity"]["value"] = enough_liquidity
+        plugin.options["ideal-ratio"]["value"] = ideal_ratio
+        failed_pairs = []
+        success = 0
+        fee_spent = 0
+        while True:
+            result = maybe_rebalance_once(plugin, failed_pairs)
+            if result["success"] == False:
```

Comment:
Nit: flake8 `E712 comparison to False should be 'if cond is False:' or 'if not cond:'`